[BEAM-1045] IOChannelFactory: removes toPath() and resolve(), and rely on URI for paths resolving.

peihe opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

…o tighten the interface.

peihe commented:
R: @jkff

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,97 @@
+   * absolute paths, then this method returns a path that starts with the last absolute path
+   * in {@code others} joined with the remaining paths.
+   */
+  public static String resolveAgainstDirectory(String directory, String... others) {
```
Why the decision to make this class operate on String rather than URI?

jkff replied:
Why do you need "others" to be an array rather than a single element?

jkff replied:
Oh, I see. However, it seems like an easy-to-misuse API: the new FileSystems API will be phrased in terms of URIs, and people will need to remember that despite that, URI.resolve() does the wrong thing, and they should use PathUtils.resolve() on uri.toString() instead. Is there any way to address this?

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  /**
+   * Returns the directory portion of the {@code path}, which includes all but the file name.
```
Document what does this return if the path ends with a / ?
What does this return if this is a path something like "gs://bucket"?
What does this return for ".." or "." or "~" ? Are these even allowed?
etc.
In general, examples would be very helpful in docs to the methods of this whole class. There was a lot of ambiguity in docs of IOChannelUtils and since we're revamping the whole thing it'd be great to also do comments properly.

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,97 @@
+   *         or an empty {@code String} if this path is "/" or empty,
+   *         or {@code null} if this path is {@code null}.
+   */
+  public static String getFileName(String path) {
```
Same questions

jkff commented on sdks/java/core/src/test/java/org/apache/beam/sdk/util/FileIOChannelFactoryTest.java:
```diff
@@ -136,8 +136,11 @@ public void testMatchNone() throws Exception {
 
-    // Windows doesn't like resolving paths with * in them, so the * is appended after resolve.
-    assertThat(factory.match(factory.resolve(temporaryFolder.getRoot().getPath(), "b") + "*"),
+    // Windows doesn't like resolving paths with * in them,
```
Is this still true for the new implementation? (also in other places in this file)

jkff commented on sdks/java/core/src/test/java/org/apache/beam/sdk/util/PathUtilsTest.java:
```diff
@@ -0,0 +1,96 @@
+/**
+ * Unit tests for {@link PathUtils}.
+ */
+public class PathUtilsTest {
```
All tests here are for a local-file path. Add tests for GCS- or S3-like paths, with buckets and stuff?

Also, I think the path resolution functions should work on non-existing files too, so you don't need an actual TemporaryFolder, you can hard-code the path of a fictional folder ("/tmp/foo" for example). The tests will probably get much cleaner since you'll drop all those tmpFolder.getRoot().toPath() calls and so on.

jkff replied:
Please add more non-trivial examples - specifically for the cases where it would be non-trivial for a human to figure out what should be the right answer. Handling the common cases is easy; a big part of the job of a class for working with filenames is defining how the corner cases are handled, in a useful way.

jkff commented on sdks/java/core/src/test/java/org/apache/beam/sdk/util/PathUtilsTest.java:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  @Test
+  public void testResolveOtherIsFullPath() throws Exception {
```
By "Full" you mean "Absolute"?

peihe commented:
PTAL

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,97 @@
+  private static final String URI_DELIMITER = "/";
+
+  /**
+   * Resolve multiple {@code others} against the {@code directory} sequentially.
```
Phrasing cleanup:

Resolves multiple path fragments against the given directory sequentially.

Unlike URI#resolve, this function appends relative path fragments rather than replaces them, e.g. (give example).

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,97 @@
+   * <p>Unlike {@link URI#resolve}, {@link #resolveAgainstDirectory} includes the last segment of
+   * the path.
+   *
+   * <p>Empty paths in {@code others} are ignored. If {@code others} contains one or more
```
Why allow empty paths at all?

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  private static String resolveAgainstDirectory(String directory, String other) {
+    if (Strings.isNullOrEmpty(other)) {
```
Why allow nulls?

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,97 @@
+   *
+   * @return a string representing the name of the file or directory,
+   *         or an empty {@code String} if this path is "/" or empty,
+   *         or {@code null} if this path is {@code null}.
```
Why allow nulls?

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,97 @@
+   *
+   * @return a string representing the directory portion of the {@code path},
+   *         or the {@code path} itself, if it is already a directory,
+   *         or an empty {@link String}, if {@code path} is "~", ".", "..".
```
I did not mean to ask for specifically documenting `~`, `.` and `..` - I would like to see the documentation describe more concretely the general rule by which it operates in all such ambiguous cases, because these are not the only ambiguous cases. And illustrate the rule by examples.

E.g. what will it return for "gs://bucket"? For "gs://"? For "file:///" ? For "/foo//bar"? For "/foo/bar/../" ? For "/foo/bar/.."? For other cases I haven't thought of?

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,97 @@
+   * the root in the directory hierarchy.
+   *
+   * @return a string representing the name of the file or directory,
+   *         or the {@code path} itself, if {@code path} is "~", ".", "..",
```
Same questions as to the previous method (ambiguities, disallowing nulls etc)

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,97 @@
+   * @return a string representing the name of the file or directory,
+   *         or the {@code path} itself, if {@code path} is "~", ".", "..",
+   *         or an empty {@link String} if this path is "/" or empty,
+   *         or {@code null} if this path is {@code null}.
```
This comment about nulls is incorrect, the code throws on null.

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,97 @@
+    checkNotNull(path, "path");
+    if (path.isEmpty() || path.equals(URI_DELIMITER)) {
+      return "";
+    } else if (path.endsWith(URI_DELIMITER)) {
```
I really don't like this hand-crafted string manipulation - it looks naive, in the sense that there are probably corner cases it doesn't cover. Are there any utilities we can reuse?

peihe commented:
PTAL

You are right that there are corners cases. I added more tests and input validations, also fixed few bugs.
(And, I did't find any URI library with similar functionalities and licenses that we can use.)

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java:
```diff
@@ -313,27 +315,24 @@ public FileBasedWriteOperation(FileBasedSink<T> sink) {
-        } catch (IOException e) {
-          throw new RuntimeException(e);
-        }
+        URI baseOutputUri = URI.create(baseOutputFilename).normalize();
```
* This fragment looks like something that should a single call to PathUtils - AFAIU the whole point of PathUtils is to free clients from this error-prone string and URI manipulation.
* I don't understand what's going on here in general:
** does URI.create() work for Windows paths (e.g. "c:\foo bar\blah")?
** What's the difference between baseOutputUri and pathUri?
** What does the self-relativize in "pathUri.resolve("").relativize(pathUri)" accomplish?
** Why do we hand-concatenate paths at line 332 instead of using .resolve()?
* Line 323 says we shouldn't call Instant.now() inside .apply(), but at line 334 we're calling it - I'm surprised there's no test for that, can you add one?

peihe replied:
1.
My worries about adding this logic to PathUtils is that people might misuse it.
I think PathUtils should only contain common URI manipulation. And, they need to be simple and easy to understand to avoid being misused.
The logic here is not simple, and the client should own it: understand the URI manipulation, and verify the code with sufficient tests.

2. 
- non-URI paths needs to be converted before reaching to here.
- pathUri doesn't contains scheme, authority, query and fragment.
- self-relativize is to get the last component of pathUri.
"/home/output" resolve("") returns "/home/"
"/home/" relativize("/home/output") returns "output"
- line 332 is constructing the directory name. It is not path resolving (append /).

3.added test for now.

jkff replied:
1 - ok, I'm convinced.

2. "non-URI paths need to be converted" - which code does that? In other words: does the current PR change user-facing behavior of FileBasedSink in a backward-incompatible way on Windows, i.e. will it break existing user pipelines? If yes, please fix that; if no, please explain why not, because if I'm understanding correctly the answer currently is yes.

On bullets 2,3 - please put them in comments in code - I asked about them because they are unclear from the code; and ideally they should be.

peihe replied:
done

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,101 @@
+   * @throws IllegalArgumentException if others contains {@link URI} query or fragment components.
+   */
+  public static String resolveAgainstDirectory(
+      @Nonnull String directory, @Nonnull String other, @Nonnull String... others) {
```
Annotating as `Nonnull` is unnecessary - everything is non-null unless annotated as `Nullable`. This is the style followed by the rest of Beam codebase.

jkff commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/PathUtils.java:
```diff
@@ -0,0 +1,101 @@
+
+    String path = directory.getPath();
+    if (!Strings.isNullOrEmpty(directory.getAuthority()) && Strings.isNullOrEmpty(path)) {
+      // Workaround of [BEAM-1174]: path needs to be absolute if the authority exists.
```
Insert the full link https://issues.apache.org/jira/browse/BEAM-1174

jkff commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileBasedSinkTest.java:
```diff
@@ -417,6 +419,89 @@ public void testGenerateOutputFilenamesWithoutExtension() {
+            .apply("/home/output/"),
+        Matchers.startsWith("/home/output/temp-beam--"));
+
+    // Tests for Windows OS path in URI format.
```
What about Windows OS path in non-URI format: it's currently allowed.

jkff replied:
By "currently" do you mean "at Beam HEAD", or "at the current PR"?

peihe replied:
I mean "at Beam HEAD", see https://issues.apache.org/jira/browse/BEAM-1045

This PR removes toPath() and resolve(). It fixed most of BEAM-1045. (still need to fix match())
I updated the PR title to refer to BEAM-1045

jkff commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileBasedSinkTest.java:
```diff
@@ -417,6 +419,89 @@ public void testGenerateOutputFilenamesWithoutExtension() {
+  public void testBuildTemporaryDirectoryName() {
+    // Tests for local files without the scheme.
+    assertThat(
+        new FileBasedWriteOperation.TemporaryDirectoryBuilder()
```
Extract `new FileBasedWriteOperation.TemporaryDirectoryBuilder()` into a local variable and reformat?

jkff commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/FileBasedSinkTest.java:
```diff
@@ -417,6 +419,89 @@ public void testGenerateOutputFilenamesWithoutExtension() {
+    assertThat(
+        new FileBasedWriteOperation.TemporaryDirectoryBuilder()
+            .apply("/home/output/"),
+        Matchers.startsWith("/home/output/temp-beam--"));
```
It might be good to special-case this case where there is no base filename, to avoid producing the confusing `--` - it looks like a bug even though it isn't.

jkff commented on sdks/java/core/src/test/java/org/apache/beam/sdk/util/PathUtilsTest.java:
```diff
@@ -0,0 +1,149 @@
+    // Tests absolute path.
+    assertEquals(
+        "/root/tmp/aa",
+        PathUtils.resolveAgainstDirectory("/root/tmp/aa", "/root/tmp/aa"));
```
Why is the original and resolved path the same?

jkff commented on sdks/java/core/src/test/java/org/apache/beam/sdk/util/PathUtilsTest.java:
```diff
@@ -0,0 +1,149 @@
+    // Tests query and fragment.
+    assertEquals(
+        "file:/home/output/aa/bb",
+        PathUtils.resolveAgainstDirectory("file:/home/output?query#fragment", "aa", "bb"));
```
It seems like we're always dropping query and fragment. Why even allow them?

jkff replied:
The reason to disallow query and fragment is that we're dropping them - if the user specifies them, they probably expect that we'll preserve this data, but we won't - dropping data should not be done silently.

peihe replied:
done
checked in resolveAgainstDirectory()

peihe commented:
PTAL

jkff commented:
Is this ready for another round?... Please get Jenkins green.

peihe commented:
Updated the PR, waiting for tests.
I did:
Rebased PR to resolve conflicts
Addressed feedback in commit ffe0238
Merged PathUtils into FileSystems in commit f11f9f8

peihe commented:
PTAL

tests passed

peihe commented:
Close this PR, and resolve() is done in https://github.com/apache/beam/pull/1900

Thanks for Eugene for reviewing. Many of the feedback are adopted in the PR above.
